---
title: "TSRchitect"
author: "R. Taylor Raborn"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

# TSRchitect Vignette

## R. Taylor Raborn
## Volker P. Brendel
## Department of Biology, Indiana University

### 16 January 2017

TSRchitect is an R package for analyzing diverse types of high-throughput transcription start site (TSS) profiling datasets. TSRchitect can handle TSS profiling experiments that contain either single-end orb paired-end sequence. In recent years, large-scale TSS profiling data has characterized the landscape of transcription initiation at high resolution, identifying promoter architecture in a number of eukaryotic model systems, including human, mouse, fruit fly and worm.

Examples of TSS profiling data types that TSRchitect is capable of handling are as follows:

- CAGE (Cap Analysis of Gene Expression) [Single-end]
- PEAT (Paired-end Analysis of Transcription) [Paired-end]
- RAMPAGE (RNA Annotation and Mapping of Promoters for Analysis of Gene Expression) [Paired-end]
- TSS-seq [Single-end]

TSRchitect provides the capability to efficiently identify putative promoters---which we call transcription start regions (TSRs)---from TSS profiling experiments. TSRchitect possesses the flexibility to accomodate datasets, including biological replicates and multiple tissues/conditions, that were generated in a variety of model organisms and genome assemblies, requiring only aligned TSS profiling information (in BAM format) as the initial input. To aid the downstream analysis of identified promoters, TSRchitect calcualtes a variety of TSR properties that have been shown to be associated with promoter architecture, including TSR activity, width and the Shape Index (SI). Finally, TSRchitect's output is compatible with popular differential expression analyses such as edgeR, assisting in downstream analysis to identify TSRs that are differentially active in one sample versus another. In addition to this vignette, the TSRchitect Reference Manual is available as part of the package's online documentation.

## Identifying promoters from RAMPAGE data derived from two human cell lines.

RAMPAGE is a TSS profiling method that identifies promoters at large-scale using a cap-based library construction method that is adapted for paired-end sequencing. Developed recently by Batut and Gingeras (2013), it has become a popular method for promoter identification and is currently part of the data compredium in the latest edition of the ENCODE project.

In this example we will process RAMPAGE data derived from two immortalized human cell lines with TSRchitect. The experiments selected for this vignette are part of the ENCODE project and is publically available online at the [ENCODE Experiment matrix](https://www.encodeproject.org/matrix/?type=Experiment). The two samples come from HT1080 cells, which is a well-characterized fibrosarcoma cell line, and NCI-H460 cells, which are derived from a large cell lung carcinoma in a male patient. 

To begin, we must first download the RAMPAGE datasets (which were aligned to GRCh38 and are in BAM format) to our local system. Do accomplish this we will utilize the "ENCODExplorer" package, which is part of the Bioconductor suite. More information on ENCODExplorer package, including information in installation, can be at the following link: https://www.bioconductor.org/packages/release/bioc/html/ENCODExplorer.html 

```{r eval=FALSE}
library(ENCODExplorer)
data(encode_df, package = "ENCODExplorer")
datasets <- fuzzySearch(searchTerm = c("ENCFF474YPI", "ENCFF242UWH", "ENCFF214GWH", "ENCFF265SGZ"), database = encode_df, filterVector = c("file_accession"), multipleTerm = TRUE)
downloadEncode(datasets, df=encode_df, format="bam")
```

Once the above steps are complete, please move the files into a new folder called `bam_files/` inside the demo/ folder.

We'll also need to download a human gene annotation, which will be important a little later.

Please download and uncompress the annoation file and place it in the `vignettes/` folder.

The file is found here:
ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gff3.gz

Now that we have our input and annotation files prepared, we can load TSRchitect into our workspace, along with another library required for parallelization:

```{r eval=FALSE}
library(TSRchitect) #loading TSRchitect
library(doParallel) #loading the doParallel package, which is requried for parallel processing
```

To implement parallelization, we'll need to set up a cluster of 6 nodes first. (Please adjust the number of clusters as you desire and your compute permits):

```{r eval=FALSE}
mycl <- makeCluster(6,type="FORK",outfile="")
registerDoParallel(mycl)
```

Next we'll create the dedicated S4 object---called the tssObj---that TSRchitect's functions will be performed on.

```{r eval=FALSE}
initializeExp(expTitle ="Human RAMPAGE", experimentName="Hs_RAMPAGE", expDir="vignettes/bam_files/", isPairedEnd=TRUE) 
```

This will create an object called `Hs_RAMPAGE`, in your workspace. We specified the directory "vignette_files" as the location of the bam files to be imported, and have given it the rather general title "Human RAMPAGE". Before we continue, let's take a look look at it to make sure it appears:

```{r eval=FALSE}
ls()
```
Unless you have been doing other work in your R console, that command should just return our tssObj S4 object:

[1] "Hs_RAMPAGE"

Now we need to provide the sample names and specify which samples are biological replicates. In this case we are working with 4 total datasets and 2 samples in duplicate, and we provide the names as follows:

```{r eval=FALSE}
setSampleID(Hs_RAMPAGE, c("HT1080-1","HT1080-2","H460-1", "H460-2"), c(1,1,2,2))
```

Now that we have completed the above commands, we can proceed with importing the bam files using the function `importBam`. This will attach GenomicAlignments objects (representing the four bam files in this example) to your tssObject. 

```{r eval=FALSE}
importBam(Hs_RAMPAGE)

... importBam ...

Importing paired-end reads ...

TSS data were specified to be paired-end read alignments.
Beginning import of  4  bam files ...
```

You will receive a message on your console when all four files have been imported:

```{r eval=FALSE}
Done. Alignment data from  4  bam files have been attached to tssObject
object " Hs_RAMPAGE ".
 Done.
-------------------------------------------------------------------------------- 
```

Now that the alignment files have been imported and attached to our tssObj S4 object, we continue by computing the TSSs from the bam files.

```{r eval=FALSE}
bamToTSS(experimentName=Hs_RAMPAGE)
```

Next we will calculate the abundance of each tag in our TSS datasets.

```{r eval=FALSE}
processTSS(experimentName=Hs_RAMPAGE, parallel=TRUE, tssSet="all", writeTable=TRUE)
```

Since we specified 'writeTable=TRUE', files (entitled "TSSset-1.txt" to "TSSset-4.txt") containing TSS abundance will be written into your working directory. 

Now that we have calculated the abundance for each TSS in the previous step we can calculate TSRs (promoters) on each of the 4 separate datasets. We have selected a tagCount threshold of 25 tags in order for a TSS to be considered. As with the previous step, we selected 'writeTable=TRUE' and therefore you will find the file ("TSRset-1" to "TSRset-4") written to your working directory.

```{r eval=FALSE}
determineTSR(experimentName = Hs_RAMPAGE, parallel = TRUE, tsrSetType = "replicates", tssSet="all", tagCountThreshold=25, clustDist=20, writeTable=TRUE)
```

To calculate TSRs from each sample (as opposed to each replicate) we need to combine our replicate data. This will be done using the identifiers we specified on our tssObj S4 object using the function initializeExp().

```{r eval=FALSE}
mergeSampleData(experimentName=Hs_RAMPAGE)
```

Having combined the TSS abundance of replicate data into samples, we next proceed with identifying TSRs for the two samples individually. We specify this with 'tsrSetType="merge"'.

```{r eval=FALSE}
determineTSR(experimentName = Hs_RAMPAGE, parallel = TRUE, tsrSetType = "merged", tssSet="all", tagCountThreshold=40, clustDist=20, writeTable=TRUE)
stopCluster(mycl) #ending the parallel session we set up earlier in the vignette.
```

Now we will calculate the number of tags from each experiment within the combined set of TSRs.

```{r eval=FALSE}
addTagCountsToTSR(experimentName= Hs_RAMPAGE, tsrSetType= "merged", tsrSet=3, tagCountThreshold= 40, writeTable= TRUE)
```

### Associating identified TSRs with gene annotations

Now that identifying TSRs are complete an obvious and biologically useful step is to determine which TSRs are adjacent to annotated genes, and to retrieve the appropriate gene IDs. Before doing this, it is imperative to select an annotation file that was generated for the assembly the reads were aligned to. Please make sure you have already downloaded the annotation file linked to at the beginning of this vignette.

```{r eval=FALSE}
importAnnotation(experimentName = Hs_RAMPAGE, fileType = "gff3", annotFile = "vignettes/gencode.v19.annotation.gff3")
```
Now we will associate the gene annotation to the TSRs within our two merged samples. We selected the feature 'transcript' from the Gencode annotation. 

```{r eval=FALSE}
addAnnotationToTSR(experimentName = Hs_RAMPAGE, tsrSetType="merged", tsrSet=1, upstreamDist=1000, downstreamDist=200, feature="transcript", featureColumnID="ID", writeTable=TRUE)

addAnnotationToTSR(experimentName = Hs_RAMPAGE, tsrSetType="merged", tsrSet=2, upstreamDist=1000, downstreamDist=200, feature="transcript", featureColumnID="ID", writeTable=TRUE)
```

Finally, we will repeat the two commands above, only associating the gene annotation to the "combined" set of TSRs, which is found in the 3rd position on the @tsrDataMerged slot.

```{r eval=FALSE}
addAnnotationToTSR(experimentName = Hs_RAMPAGE, tsrSetType="merged", tsrSet=3, upstreamDist=1000, downstreamDist=200, feature="transcript", featureColumnID="ID", writeTable=TRUE)
```

Let's briefly look at our sets of identified TSRs.

```{r eval=FALSE}
Hs_RAMPAGE@tsrDataMerged[[1]] -> HT1080.tsrs
dim(HT1080.tsrs)
```

[1] 18040     8 

```{r eval=FALSE}
Hs_RAMPAGE@tsrDataMerged[[2]] -> H460.tsrs
dim(H460)
```

[1] 15904     8 

```{r eval=FALSE}
Hs_RAMPAGE@tsrDataMerged[[3]] -> combined.tsrs
dim(combined.tsrs)
```

[1] 22750    12

We see that there are 22750 TSRs identified in the combined set, and 18040 and 15904 TSRs in the HT1080 and H460 samples, respectively. We also notice that there are 5 additional columns in the combined set. This is due to us previouly adding tag counts to the combined set of tSRs using addTagCountsToTSR, something we did not do in this vignette for the two samples.

You now have a complete set of TSS and TSR data attached to your tssObj S4 object, in addition the tables that were already written to your working directory.

These data can be accessed from the tssObj simply. For example, if we want to copy the TSR annotations from the HT1080 cells onto our R workspace, we could type the following command:

```{r eval=FALSE}
Hs_RAMPAGE@tsrDataMerged[[1]] -> HT1080.tsrs
head(HT1080.tsrs) #printing the first few rows of the table to our console.
```

This covers the main features of TSRchitect at present. If we wish to save our tssObj and return to our work later, simply we simply type the following, which will write an R binary to your working directory.

```{r eval=FALSE}
save(Hs_RAMPAGE, file="Hs_RAMPAGE.RData")
```

