<!--
vignette: >
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteIndexEntry{TSRchitect}
    %\VignettePackage{TSRchitect}
    %\VignetteIndexExtry{TSRchitect}
    \usepackage[utf8]{inputenc}
-->
```{r setup, include=FALSE}
library(knitr)
```

## TSRchitect Package Vignette

### R. Taylor Raborn, Indiana University

### 13 January 2017

TSRchitect is an R package for analyzing diverse types of high-throughput transcription start site (TSS) profiling datasets. TSRchitect can handle TSS profiling experiments that contain either single-end orb paired-end sequence. In recent years, large-scale TSS profiling data has characterized the landscape of transcription initiation at high resolution, identifying promoter architecture in a number of eukaryotic model systems, including human, mouse, fruit fly and worm.

Examples of TSS profiling data types that TSRchitect is capable of handling are as follows:

- CAGE (Cap Analysis of Gene Expression) [Single-end]
- PEAT (Paired-end Analysis of Transcription) [Paired-end]
- RAMPAGE (RNA Annotation and Mapping of Promoters for Analysis of Gene Expression) [Paired-end]
- TSS-seq [Single-end]

TSRchitect provides the capability to efficiently identify putative promoters---which we call transcription start regions (TSRs)---from TSS profiling experiments. TSRchitect possesses the flexibility to accomodate datasets, including biological replicates and multiple tissues/conditions, that were generated in a variety of model organisms and genome assemblies, requiring only aligned TSS profiling information (in BAM format) as the initial input. To aid the downstream analysis of identified promoters, TSRchitect calcualtes a variety of TSR properties that have been shown to be associated with promoter architecture, including TSR activity, width and the Shape Index (SI). Finally, TSRchitect's output is compatible with popular differential expression analyses such as edgeR, assisting in downstream analysis to identify TSRs that are differentially active in one sample versus another. In addition to this vignette, the TSRchitect Reference Manual is available as part of the package's online documentation.

To install TSRchitect, please type the following from an R console:

```{r eval=FALSE}
library(devtools) #loads devtools (please install this package if you haven't already done so)
install_github("brendelgroup/TSRchitect")
```

## Example 1: Identifying promoters from RAMPAGE data derived from two human cell lines.

RAMPAGE is a TSS profiling method that identifies promoters at large-scale using a cap-based library construction method that is adapted for paired-end sequencing. Developed recently by Batut and Gingeras (2013), it has become a popular method for promoter identification and is currently part of the data compredium in the latest edition of the ENCODE project.

In this example we will process RAMPAGE data derived from two immortalized human cell lines with TSRchitect. The experiments selected for this vignette are part of the ENCODE project and is publically available online at the [ENCODE Experiment matrix](https://www.encodeproject.org/matrix/?type=Experiment). The two samples come from HT1080 cells, which is a well-characterized fibrosarcoma cell line, and NCI-H460 cells, which are derived from a male large cell lung carcinoma. 

Before we begin, please download the RAMPAGE datasets (which were aligned to GRCh38 and are in BAM format) from the following locations:

HT1080:
-[ENCFF474YPI-rep1](https://www.encodeproject.org/files/ENCFF474YPI/@@download/ENCFF474YPI.bam)
-[ENCFF242UWH-rep2](https://www.encodeproject.org/files/ENCFF242UWH/@@download/ENCFF242UWH.bam)

NCI-H460:
-[NCI-H460-rep1](https://www.encodeproject.org/files/ENCFF214GWH/@@download/ENCFF214GWH.bam)
-[NCI-H460-rep2](https://www.encodeproject.org/files/ENCFF265SGZ/@@download/ENCFF265SGZ.bam)

Please download these to your machine and place them in a new folder called ```bam_files/``` inside the demo/ folder.

Now that we have our input files prepared, we can load TSRchitect into our workspace, along with another library required for parallelization:

```{r eval=FALSE}
library(TSRchitect) #loading TSRchitect
library(doParallel) #loading the doParallel package, which is requried for parallel processing
```

To implement parallelization, we'll need to set up a cluster of 6 nodes first. (Please adjust the number of clusters as you desire and your compute permits):

```{r eval=FALSE}
mycl <- makeCluster(6,type="FORK",outfile="")
registerDoParallel(mycl)
```

Next we'll create the dedicated S4 object---called the tssObj---that TSRchitect's functions will be performed on.

```{r eval=FALSE}
initializeExp("Human RAMPAGE lung", "Hs_lung_RAMPAGE", "/scratch/rtraborn/RAMPAGE_ENCODE/vignette_files", isPairedEnd=TRUE) 
```

This will create an object, `Hs_lung_RAMPAGE`, in your working directory. Let's take a look at our workspace to see it there:

```{r eval=FALSE}
ls()
```


Feel free to draw beautiful plots and write math $P(X>x)=\alpha/2$.

```{r}
n=300; set.seed(123)
par(mar=c(4,4,.1,.1))
plot(rnorm(n), rnorm(n), pch=21, cex=5*runif(n), col='white', bg='gray')
```

You can use your own CSS file instead of the built-in style in the **markdown** package -- just set the option `markdown.HTML.stylesheet`, e.g.

```{r css, eval=FALSE}
options(markdown.HTML.stylesheet = 'path/to/a/custom/style.css')
```
