<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="R. Taylor Raborn" />

<meta name="date" content="2023-01-02" />

<title>TSRchitect User’s Guide</title>

<script src="data:application/javascript;base64,Ly8gUGFuZG9jIDIuOSBhZGRzIGF0dHJpYnV0ZXMgb24gYm90aCBoZWFkZXIgYW5kIGRpdi4gV2UgcmVtb3ZlIHRoZSBmb3JtZXIgKHRvCi8vIGJlIGNvbXBhdGlibGUgd2l0aCB0aGUgYmVoYXZpb3Igb2YgUGFuZG9jIDwgMi44KS4KZG9jdW1lbnQuYWRkRXZlbnRMaXN0ZW5lcignRE9NQ29udGVudExvYWRlZCcsIGZ1bmN0aW9uKGUpIHsKICB2YXIgaHMgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yQWxsKCJkaXYuc2VjdGlvbltjbGFzcyo9J2xldmVsJ10gPiA6Zmlyc3QtY2hpbGQiKTsKICB2YXIgaSwgaCwgYTsKICBmb3IgKGkgPSAwOyBpIDwgaHMubGVuZ3RoOyBpKyspIHsKICAgIGggPSBoc1tpXTsKICAgIGlmICghL15oWzEtNl0kL2kudGVzdChoLnRhZ05hbWUpKSBjb250aW51ZTsgIC8vIGl0IHNob3VsZCBiZSBhIGhlYWRlciBoMS1oNgogICAgYSA9IGguYXR0cmlidXRlczsKICAgIHdoaWxlIChhLmxlbmd0aCA+IDApIGgucmVtb3ZlQXR0cmlidXRlKGFbMF0ubmFtZSk7CiAgfQp9KTsK"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="data:text/css,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">TSRchitect User’s Guide</h1>
<h4 class="author">R. Taylor Raborn</h4>
<h4 class="date">2023-01-02</h4>



<div id="tsrchitect-users-guide" class="section level1">
<h1>TSRchitect User’s Guide</h1>
<div id="r.-taylor-raborn-and-volker-p.-brendel" class="section level2">
<h2>R. Taylor Raborn and Volker P. Brendel</h2>
</div>
<div id="department-of-biology-indiana-university" class="section level2">
<h2>Department of Biology, Indiana University</h2>
</div>
<div id="first-edition-25-january-2017" class="section level2">
<h2>First edition 25 January 2017</h2>
</div>
<div id="last-updated-2-january-2023" class="section level2">
<h2>Last updated 2 January 2023</h2>
<p>TSRchitect is an R package for analyzing diverse types of high-throughput transcription start site (TSS) profiling datasets. In recent years, large-scale TSS profiling data has characterized the landscape of transcription initiation at high resolution, identifying promoter architecture in a number of eukaryotic model systems, including human, mouse, fruit fly and worm. TSRchitect can handle TSS profiling experiments that contain either single-end or paired-end sequence reads.</p>
<p>Examples of TSS profiling data types that TSRchitect is capable of handling are:</p>
<ul>
<li>CAGE (Cap Analysis of Gene Expression) [Single-end]</li>
<li>PEAT (Paired-end Analysis of Transcription) [Paired-end]</li>
<li>RAMPAGE (RNA Annotation and Mapping of Promoters for Analysis of Gene Expression) [Paired-end]</li>
<li>CapSeq [Single-end]</li>
<li>TSS-seq [Single-end]</li>
<li>STRIPE-seq (Survey of Transcription Initiation at Promoters and Enhancers) [Single-end or Paired-end]</li>
</ul>
<p>TSRchitect provides the capability to efficiently identify putative promoters—which we call transcription start regions (TSRs)—from TSS profiling experiments. TSRchitect can accommodate multiple datasets, including biological replicates and multiple tissues/conditions, that were generated in a variety of model organisms and genome assemblies, requiring only aligned TSS profiling information (in BAM format) as the initial input. To aid the downstream analysis of identified promoters, TSRchitect calculates a variety of TSR properties that have been shown to be associated with promoter architecture, including TSR activity, width, Shape Index (SI), modified Shape Index (mSI) and torque. Finally, TSRchitect’s output is compatible with popular differential expression software such as <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>, assisting in downstream analysis to identify TSRs that are differentially active in one sample versus another. In addition to this vignette, the TSRchitect Reference Manual is available as part of the package’s online documentation. Please note that this document will guide you through a few examples of analysis; if you’re intrested in a detailed discussion of individual functions, please see the TSRchitect Reference Manual, which is found in <code>man/TSRchitect.pdf</code> and online <a href="https://bioconductor.org/packages/release/bioc/manuals/TSRchitect/man/TSRchitect.pdf">here</a>.</p>
</div>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>In your current directory, create subdirectories as follows on the command line.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> downloads</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> downloads</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> HsRAMPAGEbam</span></code></pre></div>
<p>Now that this is complete, we proceed with the first of the three examples contained in this vignette.</p>
</div>
<div id="example-1-identifying-promoters-from-rampage-data-derived-from-two-human-cell-lines." class="section level2">
<h2>Example 1: Identifying promoters from RAMPAGE data derived from two human cell lines.</h2>
<p>RAMPAGE is a TSS profiling method that identifies promoters at large-scale using a cap-based library construction method that is adapted for paired-end sequencing. Developed recently by Batut and Gingeras (2013), it has become a popular method for promoter identification and is currently part of the data compredium in the latest edition of the ENCODE project.</p>
<p>In this example we will process RAMPAGE data derived from two immortalized human cell lines with TSRchitect. The experiments selected for this vignette are part of the ENCODE project and are publically available online at the <a href="https://www.encodeproject.org/matrix/?type=Experiment">ENCODE Experiment matrix</a>. The two samples come from HT1080 cells, which is a well-characterized fibrosarcoma cell line, and NCI-H460 cells, which are derived from a large cell lung carcinoma in a male patient.</p>
<p>To begin, we must first download the RAMPAGE datasets (which were aligned to GRCh38 and are in BAM format) to our local system. To accomplish this we will utilize the “ENCODExplorer” package, which is part of the Bioconductor suite. More information on ENCODExplorer package can be found at the following link: <a href="https://www.bioconductor.org/packages/release/bioc/html/ENCODExplorer.html" class="uri">https://www.bioconductor.org/packages/release/bioc/html/ENCODExplorer.html</a>.</p>
<p>Now we can proceed (in the R console) with downloading the data:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Downloading the files:</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">R_USER_CACHE_DIR=</span><span class="st">&quot;./tmpRcache&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ENCODExplorer)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>encode_df <span class="ot">&lt;-</span> <span class="fu">get_encode_df</span>()</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>datasets <span class="ot">&lt;-</span> <span class="fu">fuzzySearch</span>(<span class="at">searchTerm=</span><span class="fu">c</span>(<span class="st">&quot;ENCFF214GWH&quot;</span>, <span class="st">&quot;ENCFF265SGZ&quot;</span>, <span class="st">&quot;ENCFF242UWH&quot;</span>, <span class="sc">/</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;ENCFF348EKW&quot;</span>), <span class="at">database=</span>encode_df, <span class="at">filterVector=</span><span class="fu">c</span>(<span class="st">&quot;file_accession&quot;</span>), <span class="sc">/</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">multipleTerm=</span><span class="cn">TRUE</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">downloadEncode</span>(datasets, <span class="at">df=</span>encode_df, <span class="at">format=</span><span class="st">&quot;bam&quot;</span>)</span></code></pre></div>
<p>Once the above steps are complete, we will move the files into the subdirectory <code>downloads/</code> to keep a record of data provenance. To use TSRchitect in this example, we coveniently use symbolic links to give more intuitive names to the datasets (as well as ordering them in proper order; see below):</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#The following are to be executed in your shell terminal:</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.bam downloads</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> downloads</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> HsRAMPAGEbam</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF214GWH.bam H460-rep1.bam</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF265SGZ.bam H460-rep2.bam</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF242UWH.bam HT1080-rep1.bam</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF348EKW.bam HT1080-rep2.bam</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre></div>
<p>Now that we have our input files prepared, we can load TSRchitect into our R workspace:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#loading TSRchitect</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TSRchitect)</span></code></pre></div>
<p>Next we’ll initialize our dedicated S4 object—called the tssObject—on which TSRchitect’s functions are applied, using <code>loadTSSobj</code>. We also need to supply other information about the experiment as arguments. This will attach GenomicAlignments objects (representing the four bam files in this example) to your tssObject. Note that we must specify <code>isPairedBAM=FALSE</code> because this is single-end CAGE data.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">loadTSSobj</span>(<span class="at">experimentTitle =</span><span class="st">&quot;Human RAMPAGE&quot;</span>, <span class="at">inputDir=</span><span class="st">&quot;HsRAMPAGEbam/&quot;</span>, <span class="sc">/</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">isPairedBAM=</span><span class="cn">TRUE</span>, <span class="at">sampleNames=</span><span class="fu">c</span>(<span class="st">&quot;H460-rep1&quot;</span>, <span class="sc">/</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;H460-rep2&quot;</span>, <span class="st">&quot;HT1080-rep1&quot;</span>, <span class="st">&quot;HT1080-rep2&quot;</span>), <span class="at">replicateIDs=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span></code></pre></div>
<p>Next we need to provide the sample names and specify which samples are biological replicates. In this case we are working with 4 total datasets and 2 samples in duplicate. Please note that, because the alignments on our <code>bamData</code> slot are organized in ascending alphabetical order (as are the file names on the <code>fileNames</code> slot), we must provide our identifiers in <code>sampleNames</code> and <code>replicateIDs</code> to directly correspond to this. To check this on the tssObject <a href="http://adv-r.had.co.nz/S4.html">S4 object</a> you have created, simply check the list of .bam files as follows using one of TSRchitect’s accessor methods.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#obtaining a list of bam files loaded to the tssObject S4 object</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getFileNames</span>(<span class="at">experimentName=</span>Hs_RAMPAGE)</span></code></pre></div>
<p>Now that the alignment files have been imported and attached to our tssObject S4 object, we continue by computing the TSSs from the alignment (in this case .bam) files.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">inputToTSS</span>(<span class="at">experimentName=</span>Hs_RAMPAGE)</span></code></pre></div>
<p>Next we will calculate the abundance of each tag in our TSS datasets. TSRchitect provides the option to run this function in parallel, so for this example we chose to run this on 4 cores. Please adjust this as your compute resource permits.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">processTSS</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="sc">/</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Since we specified ‘writeTable=TRUE’, files (entitled “TSSset-1.txt” to “TSSset-4.txt”) containing TSS abundance will be written into your working directory. Now that we have calculated the abundance for each TSS in the previous step we can calculate TSRs (promoters) on each of the 4 separate datasets using the function <code>determineTSR</code>. We select a <code>tagCount</code> threshold of 25 tags in order for a TSS to be considered. The option <code>clustDist</code> is critical for the identification of TSRs and refers to the minimum distance between distinct TSRs (in other words, adjacent TSSs seperated by more than <code>clustDist</code> nucleotides will be in different TSRs). As with the previous step, we select ‘writeTable=TRUE’, and therefore we will find the output files (“TSRset-1” to “TSRset-4”) written to the working directory. As for <code>processTSS</code>, we can run this function in parallel. We have specified 4 cores.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">determineTSR</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="sc">/</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">tssSetType=</span><span class="st">&quot;replicates&quot;</span>, <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="at">tagCountThreshold=</span><span class="dv">25</span>, <span class="at">clustDist=</span><span class="dv">20</span>, <span class="sc">/</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>To calculate TSRs from each sample (as opposed to each replicate) we need to combine our replicate data. This will be done using the identifiers we specified on our tssObject S4 object using <code>loadTSSobj</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">mergeSampleData</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="at">tagCountThreshold=</span><span class="dv">25</span>)</span></code></pre></div>
<p>Having combined the TSS abundance of replicate data into samples, we next proceed with identifying TSRs for the two samples individually. We specify this with ‘tsrSetType=“merged”’.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Generating the TSRs for the merged datasets:</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">determineTSR</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="at">tssSetType=</span><span class="st">&quot;merged&quot;</span> <span class="sc">/</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="at">tagCountThreshold=</span><span class="dv">40</span>, <span class="at">clustDist=</span><span class="dv">20</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Now calculating the number of tags from each experiment within the combined set of TSRs:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">addTagCountsToTSR</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="sc">/</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">tsrSet=</span><span class="dv">3</span>, <span class="at">tagCountThreshold=</span><span class="dv">40</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<div id="associating-identified-tsrs-with-gene-annotations" class="section level3">
<h3>Associating identified TSRs with gene annotations</h3>
<p>Now that identifying TSRs are complete, an obvious and biologically useful step is to determine which TSRs are adjacent to annotated genes, and to retrieve the appropriate gene IDs. Before doing this, it is imperative to select an annotation file that was generated for the assembly to which the reads were aligned. For this example we will retrieve the appropriate annotation files from the Bioconductor package <code>AnnotationHub</code>.</p>
<p>Please install <code>AnnotationHub</code> if you haven’t already done so.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly=</span><span class="cn">TRUE</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;AnnotationHub&quot;</span>)</span></code></pre></div>
<p>In our case we need to download the <a href="https://www.gencodegenes.org/">Gencode</a> annotation. We do this in the following manner:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>hub <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">query</span>(hub, <span class="fu">c</span>(<span class="st">&quot;gencode&quot;</span>, <span class="st">&quot;gff&quot;</span>, <span class="st">&quot;human&quot;</span>))</span></code></pre></div>
<p>This reveals nine gff3 annotations from Gencode that we can choose from. We will select the full annotation, which has the identifier “AH49555”.</p>
<pre><code>AnnotationHub with 9 records
# snapshotDate(): 2017-01-05
# $dataprovider: Gencode
# $species: Homo sapiens
# $rdataclass: GRanges
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   sourceurl, sourcetype
# retrieve records with, e.g., &#39;object[[&quot;AH49554&quot;]]&#39;

            title                                                   
  AH49554 | gencode.v23.2wayconspseudos.gff3.gz                     
  AH49555 | gencode.v23.annotation.gff3.gz                          
  AH49556 | gencode.v23.basic.annotation.gff3.gz                    
  AH49557 | gencode.v23.chr_patch_hapl_scaff.annotation.gff3.gz     
  AH49558 | gencode.v23.chr_patch_hapl_scaff.basic.annotation.gff3.gz
  AH49559 | gencode.v23.long_noncoding_RNAs.gff3.gz                 
  AH49560 | gencode.v23.polyAs.gff3.gz                              
  AH49561 | gencode.v23.primary_assembly.annotation.gff3.gz         
  AH49562 | gencode.v23.tRNAs.gff3.gz
</code></pre>
<p>Using TSRchitect, We can use the function <code>importAnnotationHub</code> to import our desired annotated record and attach it our <code>tssObject</code>. We accomplish this as follows:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">importAnnotationHub</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">provider=</span><span class="st">&quot;gencode&quot;</span>, <span class="sc">/</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotType=</span><span class="st">&quot;gff3&quot;</span>, <span class="at">species=</span><span class="st">&quot;human&quot;</span>, <span class="at">annotID=</span><span class="st">&quot;AH49555&quot;</span>)</span></code></pre></div>
<p>Next, we associate the gene annotation to the TSRs within our two merged samples. We selected the feature ‘transcript’ from the Gencode annotation.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">addAnnotationToTSR</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="sc">/</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">tsrSet=</span><span class="dv">1</span>, <span class="at">upstreamDist=</span><span class="dv">1000</span>, <span class="at">downstreamDist=</span><span class="dv">200</span>, <span class="at">feature=</span><span class="st">&quot;transcript&quot;</span>, <span class="sc">/</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">featureColumnID=</span><span class="st">&quot;ID&quot;</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">addAnnotationToTSR</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="sc">/</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tsrSet=</span><span class="dv">2</span>, <span class="at">upstreamDist=</span><span class="dv">1000</span>, <span class="at">downstreamDist=</span><span class="dv">200</span>, <span class="at">feature=</span><span class="st">&quot;transcript&quot;</span>, <span class="sc">/</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">featureColumnID=</span><span class="st">&quot;ID&quot;</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Finally, we will repeat the two commands above, instead associating the gene annotation to the “combined” set of TSRs, which is found in the 3rd position on the <code>tsrDataMerged</code> slot.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>Hs_RAMPAGE <span class="ot">&lt;-</span> <span class="fu">addAnnotationToTSR</span>(<span class="at">experimentName=</span>Hs_RAMPAGE, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="sc">/</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">tsrSet=</span><span class="dv">3</span>, <span class="at">upstreamDist=</span><span class="dv">1000</span>, <span class="at">downstreamDist=</span><span class="dv">200</span>, <span class="at">feature=</span><span class="st">&quot;transcript&quot;</span>, <span class="sc">/</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">featureColumnID=</span><span class="st">&quot;ID&quot;</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Let’s briefly look at the sets of TSRchitect-identified TSRs.</p>
<p>Using the accessor methods we applied in earlier examples, let’s take a quick glance at our set of identified TSRs.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>HT1080.tsrs <span class="ot">&lt;-</span> <span class="fu">getTSRdata</span>(Hs_RAMPAGE, <span class="at">slotType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">slot=</span><span class="dv">1</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(HT1080.tsrs)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>H460.tsrs <span class="ot">&lt;-</span> <span class="fu">getTSRdata</span>(Hs_RAMPAGE, <span class="at">slotType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">slot=</span><span class="dv">2</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(H460.tsrs)</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>combined.tsrs <span class="ot">&lt;-</span> <span class="fu">getTSRdata</span>(Hs_RAMPAGE, <span class="at">slotType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">slot=</span><span class="dv">3</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(combined.tsrs)</span></code></pre></div>
<p>Let’s look at some of the tsrs we identified on our ‘combined’ set.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(combined.tsrs)</span></code></pre></div>
<p>We see that there are 22750 TSRs identified in the combined set, and 15904 and 18040 TSRs in the H460 and HT1080 samples, respectively. We also notice that there are 5 additional columns in the combined set. This is due to us previouly having added tag counts to the combined set of TSRs using `addTagCountsToTSR’, something we did not do in this vignette for the two individual samples.</p>
<p>You now have a complete set of TSS and TSR data attached to your <code>tssObject</code> S4 object, in addition the tables that were already written to your working directory.</p>
<p>To better understand our data, let’s explore some of the characteristics of the TSRs we have identified. TSRchitect calculates the Shape Index (SI) of each TSR; the SI provides a quantitative measure of TSR (and thus promoter) shape by representing the entropy of the distribution of TSSs associated with it. An SI of 2 (which is the maximum value possible) will have only a single unique TSS coordinate, whereas a TSR with a negative SI value will have an diversity of mapped 5’ ends at distinct TSS positions. We can use SI values to classify TSRs into ‘peaked’ and ‘broad’ classes, having high and low SI values, respectively. As of TSRchitect version 1.8.0, we have included the Modified Shape Index (mSI) as an additional metric defining TSR shape. The mSI is a metric of TSR shape scaled by TSS tag abundance that can return possible values from 0 to 1 inclusive, where 1 is the most peaked (i.e. a TSR with a single unique TSS position) and 0 is the most broad TSR.</p>
<p>We can visualize the shape distribution of our identified TSRs quite easily using Hadley Wickham’s <code>ggplot2</code> graphics package, as follows. For these plots we filter will out TSRs at lower abundances (100 tags/TSR) to show only the SI values from reasonably well-sampled TSRs.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>HT1080.tsrs.filtered <span class="ot">&lt;-</span> HT1080.tsrs[HT1080.tsrs<span class="sc">$</span>nTAGs <span class="sc">&gt;</span> <span class="dv">100</span>,]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>t <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(HT1080.tsrs.filtered, <span class="fu">aes</span>(tsrSI))</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>t <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">binwidth=</span><span class="fl">0.1</span>, <span class="at">fill=</span><span class="st">&quot;blue2&quot;</span>) <span class="sc">/</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Number of Tags per TSR&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Shape Index (SI)&quot;</span>)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file=</span><span class="st">&quot;HT1080_SI.png&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>H460.tsrs.filtered <span class="ot">&lt;-</span> H460.tsrs[H460.tsrs<span class="sc">$</span>nTAGs <span class="sc">&gt;</span> <span class="dv">100</span>,]</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(H460.tsrs.filtered, <span class="fu">aes</span>(tsrSI))</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">binwidth=</span><span class="fl">0.1</span>, <span class="at">fill=</span><span class="st">&quot;darkgreen&quot;</span>) <span class="sc">/</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Number of Tags per TSR&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Shape Index (SI)&quot;</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file=</span><span class="st">&quot;H460_SI.png&quot;</span>)</span></code></pre></div>
<p>From both histograms we can see that there appears to be a bimodal distribution of TSRs within our samples; the class of ‘completely peaked’ TSRs with an SI value of 2, and another distribution of more broad TSRs with SI values centered just below 0. This is consistent with two major shape classes of TSRs known in metazoans, particularily human and mouse.</p>
<p>Similarly, we could use the same code framework above to plot the distributions of TSR width (tsrWdth), number of unique TSSs per TSR (nTSSs), the mSI and torque (balance of the TSR according to tag counts; tsrTrq); to do this we would select the appropriate column in the data frame. For example: equivalent plots can generated for mSI simply by replacing the term “tsrSI” with “tsrMSI” in the two code snippets above.</p>
<p>This concludes Example 1. Should we wish to save our tssObject and return to our work later, we simply type the following, which will write an R binary to your working directory.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(Hs_RAMPAGE, <span class="at">file=</span><span class="st">&quot;Hs_RAMPAGE.RData&quot;</span>)</span></code></pre></div>
<p><strong>Important note</strong>: before you continue with another example, please move the output files generated in your working directory to a separate, dedicated folder. Otherwise some or all of the files you generate in subsequent examples will be overwritten.</p>
</div>
</div>
<div id="example-2-identifying-promoters-in-the-model-plant-a.-thaliana-using-a-peat-dataset." class="section level2">
<h2>Example 2: Identifying promoters in the model plant A. thaliana using a PEAT dataset.</h2>
<p>For our second example will process TSS profiling data from Arabidopsis root tissue. These data come from the Megraw Lab at Oregon State University as reported in Morton et al., 2014. <a href="http://www.plantcell.org/content/26/7/2746">Link to paper:</a></p>
<p>As with the previous example, first we must download the raw data. In this case we have only a single alignment file to retrieve, which is found here: <a href="https://oregonstate.app.box.com/s/3lb3spmqbiuofhbubovc1z8bfthmxh6f" class="uri">https://oregonstate.app.box.com/s/3lb3spmqbiuofhbubovc1z8bfthmxh6f</a> Once download of the file <code>peat.sorted.bam</code> is complete, please move it to subdirectory “PEATbam/”.</p>
<p>The annotation file is available from the TAIR10 database: <a href="ftp://ftp.arabidopsis.org/home/tair/Genes/TAIR10_genome_release/TAIR10_gff3/TAIR10_GFF3_genes.gff" class="uri">ftp://ftp.arabidopsis.org/home/tair/Genes/TAIR10_genome_release/TAIR10_gff3/TAIR10_GFF3_genes.gff</a>. Please move it into the downloads/ folder you have created, e.g.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>mkdir downloads</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>cd downloads</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>wget <span class="sc">--</span>content<span class="sc">-</span>disposition <span class="sc">/</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>    ftp<span class="sc">:</span><span class="er">//</span>ftp.arabidopsis.org<span class="sc">/</span>home<span class="sc">/</span>tair<span class="sc">/</span>Genes<span class="sc">/</span>TAIR10_genome_release<span class="sc">/</span>TAIR10_gff3<span class="sc">/</span>TAIR10_GFF3_genes.gff</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>cd ..</span></code></pre></div>
<p>Because there is only a single experiment, setting the sample IDs is simple in this example:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>At_PEAT <span class="ot">&lt;-</span> <span class="fu">loadTSSobj</span>(<span class="at">experimentTitle =</span><span class="st">&quot;Arabidopsis PEAT dataset&quot;</span>, <span class="at">inputDir=</span><span class="st">&quot;PEATbam&quot;</span>, <span class="sc">/</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="at">isPairedBAM=</span><span class="cn">TRUE</span>, <span class="at">sampleNames=</span><span class="fu">c</span>(<span class="st">&quot;experiment1&quot;</span>), <span class="at">replicateIDs=</span><span class="fu">c</span>(<span class="dv">1</span>))</span></code></pre></div>
<p>Now we convert the alignment data (in this case, in .bam format) into TSS coordinates:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>At_PEAT <span class="ot">&lt;-</span> <span class="fu">inputToTSS</span>(At_PEAT)</span></code></pre></div>
<p>As in the previous example, now we can calculate the tag abundance at each location using <code>processTSS</code> and the identify TSRs within the sample using <code>determineTSR</code>. Note that we do not need to use <code>mergeSampleData</code> because there is only a single sample. As there is only a single sample we set <code>n.cores=1</code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>At_PEAT <span class="ot">&lt;-</span> <span class="fu">processTSS</span>(<span class="at">experimentName=</span>At_PEAT, <span class="at">n.cores=</span><span class="dv">1</span>, <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>At_PEAT <span class="ot">&lt;-</span> <span class="fu">determineTSR</span>(<span class="at">experimentName=</span>At_PEAT, <span class="at">n.cores=</span><span class="dv">1</span>, <span class="at">tssSetType=</span><span class="st">&quot;replicates&quot;</span>, <span class="sc">/</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="at">tagCountThreshold=</span><span class="dv">25</span>, <span class="at">clustDist=</span><span class="dv">20</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<div id="associating-identified-tsrs-with-gene-annotations-1" class="section level3">
<h3>Associating identified TSRs with gene annotations</h3>
<p>We continue by associating our newly-identified TSRs with genes from the TAIR10 annotation. Note that we use different parameters for upstreamDist and downstreamDist than we did in Example 1. This is due to the high degree of compactness in the A. thaliana genome.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>At_PEAT <span class="ot">&lt;-</span> <span class="fu">importAnnotationExternal</span>(<span class="at">experimentName=</span>At_PEAT, <span class="at">fileType=</span><span class="st">&quot;gff3&quot;</span>, <span class="sc">/</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotFile=</span><span class="st">&quot;downloads/TAIR10_GFF3_genes.gff&quot;</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>At_PEAT <span class="ot">&lt;-</span> <span class="fu">addAnnotationToTSR</span>(<span class="at">experimentName=</span>At_PEAT, <span class="at">tsrSetType=</span><span class="st">&quot;replicates&quot;</span>, <span class="at">tsrSet=</span><span class="dv">1</span>, <span class="sc">/</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">upstreamDist=</span><span class="dv">500</span>, <span class="at">downstreamDist=</span><span class="dv">200</span>, <span class="at">feature=</span><span class="st">&quot;gene&quot;</span>, <span class="at">featureColumnID=</span><span class="st">&quot;ID&quot;</span>, <span class="sc">/</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Now we have a complete set of TSRs on our tssObj object. Let’s take a look at them using one of our accessor methods.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>At.tsrs <span class="ot">&lt;-</span> <span class="fu">getTSRdata</span>(At_PEAT, <span class="at">slotType=</span><span class="st">&quot;replicates&quot;</span>, <span class="at">slot=</span><span class="dv">1</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(At.tsrs)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(At.tsrs)</span></code></pre></div>
<p>We can optionally save the tssObject as we have previously.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(At_PEAT, <span class="at">file=</span><span class="st">&quot;At_PEAT_vignette.RData&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="example-3-analysis-of-cage-datasets-from-the-encode-project" class="section level2">
<h2>Example 3: Analysis of CAGE datasets from the ENCODE project</h2>
<p>As we stated in the introduction of this vignette, TSRchitect is capable of handling diverse forms of TSS profiling data. In the first two examples, we analyze two distinct paired-end datasets: RAMPAGE and PEAT, respectively. In this example we will process data from CAGE, which is the most widely-used TSS profiling method to date. We will analyze CAGE data generated in two well-characterized immortalized cell lines, MCF-7 and A549. MCF-7 cells are derived from a breast cancer tumor, and A549 originates from an adenocarinoma isolated from lung tissue. Both datasets are part of the ENCODE project, and therefore we can make use of the ENCODExplorer package that we originally introduced in Example 1.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Downloading the files:</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">R_USER_CACHE_DIR=</span><span class="st">&quot;./tmpRcache&quot;</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ENCODExplorer)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>encode_df <span class="ot">&lt;-</span> <span class="fu">get_encode_df</span>()</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>cage_data <span class="ot">&lt;-</span> <span class="fu">fuzzySearch</span>(<span class="at">searchTerm=</span><span class="fu">c</span>(<span class="st">&quot;ENCFF552BXH&quot;</span>,<span class="st">&quot;ENCFF288VTZ&quot;</span>,<span class="st">&quot;ENCFF265RSX&quot;</span>, <span class="sc">/</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;ENCFF944PCJ&quot;</span>), <span class="at">database=</span>encode_df, <span class="sc">/</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">filterVector=</span><span class="fu">c</span>(<span class="st">&quot;file_accession&quot;</span>), <span class="at">multipleTerm=</span><span class="cn">TRUE</span>)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="fu">downloadEncode</span>(cage_data, <span class="at">df=</span>encode_df, <span class="at">format=</span><span class="st">&quot;bam&quot;</span>)</span></code></pre></div>
<p>Now that the files have been downloaded, we will create symbolic links with the appropriate sample names. Please run the following commands from a linux command line:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> downloads <span class="co"># ignore if the directory already exists</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.bam downloads</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> HsCAGEbam</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> HsCAGEbam</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF265RSX.bam A549-rep1.bam</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF944PCJ.bam A549-rep2.bam</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF552BXH.bam MCF7-rep1.bam</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../downloads/ENCFF288VTZ.bam MCF7-rep2.bam</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre></div>
<p>As in Example 1, we also need to provide human gene annotation. Please download and uncompress the annoation file found at the following location [Note: you may ignore this if you have already downloaded the file from Example 1]:</p>
<p><a href="ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gff3.gz" class="uri">ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gff3.gz</a>.</p>
<p>Please move the uncompressed file into the downloads/ directory if it does not already exist there, e.g. with</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> downloads</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--content-disposition</span> ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gff3.gz</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> gencode.v19.annotation.gff3.gz</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre></div>
<p>Now we can set up the tssObject S4 object. Note that we must specify <code>isPairedBAM=FALSE</code> because this is single-end CAGE data.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># initializing the tssObject, setting the sample IDs and importing the CAGE data</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">loadTSSobj</span>(<span class="at">experimentTitle =</span><span class="st">&quot;Human CAGE&quot;</span>, <span class="at">inputDir=</span><span class="st">&quot;HsCAGEbam&quot;</span>, <span class="sc">/</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">isPairedBAM=</span><span class="cn">FALSE</span>, <span class="at">sampleNames=</span><span class="fu">c</span>(<span class="st">&quot;A549-rep1&quot;</span>,<span class="st">&quot;A549-rep2&quot;</span>, <span class="sc">/</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;MCF7-rep1&quot;</span>,<span class="st">&quot;MCF7-rep2&quot;</span>), <span class="at">replicateIDs=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>) )</span></code></pre></div>
<p>As before, it is vital to provide arguments to <code>sampleNames</code> and <code>replicateIDs</code> in the order of the files on the <code>fileNames</code> slot to they exactly correspond to the alignment files (in this case .bam) that we imported.</p>
<p>As in the prior two examples, we then extract TSS information from the attached alignment data (which was placed on slot <code>@bamData</code>):</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Converting the alignment data into TSS information and attaching it to the tssObject:</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">inputToTSS</span>(<span class="at">experimentName=</span>CAGEhuman)</span></code></pre></div>
<p>Next we must calculate the CAGE tag abundance at each TSS position, followed by identification of TSRs within our 4 replicate datasets. As in the first example, we choose to run <code>processTSS</code> and <code>determineTSR</code> in parallel on 4 cores. Please adjust this parameter as needed.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Constructing the tag count per TSS data matrix:</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">processTSS</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="sc">/</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Finding TSRs for the replicate datasets:</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">determineTSR</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="sc">/</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">tssSetType=</span><span class="st">&quot;replicates&quot;</span>, <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="at">tagCountThreshold=</span><span class="dv">25</span>, <span class="sc">/</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">clustDist=</span><span class="dv">20</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Now we merge data from replicates into their two corresponding samples.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Merging TSS data from the replicates:</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">mergeSampleData</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="at">tagCountThreshold=</span><span class="dv">25</span>) </span></code></pre></div>
<p>Once this is complete, we can complete TSR identification on the merged samples.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Finding TSRs for the merged samples and adding tag counts:</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">determineTSR</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">n.cores=</span><span class="dv">4</span>, <span class="at">tssSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="sc">/</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">tssSet=</span><span class="st">&quot;all&quot;</span>, <span class="at">tagCountThreshold=</span><span class="dv">40</span>, <span class="at">clustDist=</span><span class="dv">20</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">addTagCountsToTSR</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="sc">/</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tsrSet=</span><span class="dv">3</span>, <span class="at">tagCountThreshold=</span><span class="dv">40</span>, <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Now we need to import the annotation file and attach it to our tssObj S4 object. To do this, we will to use the same record (Gencode v. 23) that we referred to in Example 1.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">importAnnotationHub</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">provider=</span><span class="st">&quot;gencode&quot;</span>, <span class="sc">/</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotType=</span><span class="st">&quot;gff3&quot;</span>, <span class="at">species=</span><span class="st">&quot;human&quot;</span>, <span class="at">annotID=</span><span class="st">&quot;AH49555&quot;</span>)</span></code></pre></div>
<p>Next we associate the gene annotation to the TSRs within our two merged samples i) MCF7 cells and ii) A549 cells. As we did in Example 1, we select the feature ‘transcript’ from the Gencode annotation.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">addAnnotationToTSR</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">tsrSet=</span><span class="dv">1</span>, <span class="sc">/</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">upstreamDist=</span><span class="dv">1000</span>, <span class="at">downstreamDist=</span><span class="dv">200</span>, <span class="at">feature=</span><span class="st">&quot;transcript&quot;</span>, <span class="at">featureColumnID=</span><span class="st">&quot;ID&quot;</span>, <span class="sc">/</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeTable=</span><span class="cn">TRUE</span>) <span class="co">#A549 cells</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">addAnnotationToTSR</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">tsrSet=</span><span class="dv">2</span>, <span class="sc">/</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="at">upstreamDist=</span><span class="dv">1000</span>, <span class="at">downstreamDist=</span><span class="dv">200</span>, <span class="at">feature=</span><span class="st">&quot;transcript&quot;</span>, <span class="at">featureColumnID=</span><span class="st">&quot;ID&quot;</span>, <span class="sc">/</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeTable=</span><span class="cn">TRUE</span>) <span class="co">#MCF7 cells</span></span></code></pre></div>
<p>Associating the selected annotation features with the TSRs on the ‘combined’ slot:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>CAGEhuman <span class="ot">&lt;-</span> <span class="fu">addAnnotationToTSR</span>(<span class="at">experimentName=</span>CAGEhuman, <span class="at">tsrSetType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">tsrSet=</span><span class="dv">3</span>, <span class="sc">/</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">upstreamDist=</span><span class="dv">1000</span>, <span class="at">downstreamDist=</span><span class="dv">200</span>, <span class="at">feature=</span><span class="st">&quot;transcript&quot;</span>, <span class="at">featureColumnID=</span><span class="st">&quot;ID&quot;</span>, <span class="sc">/</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeTable=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Using the accessor methods we applied in earlier examples, let’s take a quick glance at our set of identified TSRs.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getTSRdata</span>(CAGEhuman, <span class="at">slotType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">slot=</span><span class="dv">1</span>) <span class="ot">-&gt;</span> MCF7.tsrs</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(MCF7.tsrs)</span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getTSRdata</span>(CAGEhuman, <span class="at">slotType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">slot=</span><span class="dv">2</span>) <span class="ot">-&gt;</span> A549.tsrs</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(A549.tsrs)</span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getTSRdata</span>(CAGEhuman, <span class="at">slotType=</span><span class="st">&quot;merged&quot;</span>, <span class="at">slot=</span><span class="dv">3</span>) <span class="ot">-&gt;</span> CAGEhuman.tsrs</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(CAGEhuman.tsrs)</span></code></pre></div>
<p>Let’s look at some of the tsrs we identified on our ‘combined’ set.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(CAGEhuman.tsrs)</span></code></pre></div>
<p>We can optionally save the tssObject for future use:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(CAGEhuman, <span class="at">file=</span><span class="st">&quot;CAGEhuman-vignette.RData&quot;</span>)</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
